libazureinit/src/health.rs

```diff
@@ -0,0 +1,248 @@
+}
+
+/// Report that provisioning has failed.
+pub async fn report_failure(
+    message: &str,
+    config: &Config,
+    url_override: Option<&str>,
+) -> Result<(), Error> {
+    tracing::info!(failure_reason=%message, "Reporting failure provisioning health");
+    _report(
+        "NotReady",
+        Some("ProvisioningFailed"),
+        Some(message),
+        &config.wireserver,
+        url_override,
+    )
+    .await
+}
+
+/// Internal helper: all of the header setup, JSON‐body construction, retry loop, etc.
+#[instrument(err, skip_all)]
+async fn _report(
+    state: &str,
+    substatus: Option<&str>,
+    description: Option<&str>,
+    cfg: &crate::config::Wireserver,
+    url_override: Option<&str>,
+) -> Result<(), Error> {
+    let body = if state == "NotReady" {
+        json!({
+            "state": state,
+            "details": {
+                "subStatus": substatus.unwrap_or("Provisioning"),
+                "description": description
+            }
+        })
+        .to_string()
+    } else {
+        json!({ "state": state }).to_string()
+    };
+
+    tracing::debug!(body=%body, "Built provisioning-health JSON");
+
+    let version = env!("CARGO_PKG_VERSION");
+    let mut headers = HeaderMap::new();
+    headers.insert(
+        USER_AGENT,
+        HeaderValue::from_str(&format!("azure-init v{}", version)).unwrap(),
+    );
+    headers.insert(
+        "x-ms-guest-agent-name",
+        HeaderValue::from_str(&format!("azure-init/{}", version)).unwrap(),
+    );
+    headers
+        .insert("content-type", HeaderValue::from_static("application/json"));
+
+    tracing::info!(?headers, "Prepared HTTP headers");
+
+    let url = url_override.unwrap_or(DEFAULT_HEALTH_URL);
+
+    let connect_timeout = Duration::from_secs_f64(cfg.connection_timeout_secs);
+    let read_timeout = Duration::from_secs_f64(cfg.read_timeout_secs);
+    let retry_for = Duration::from_secs_f64(cfg.total_retry_timeout_secs);
+
+    let client = Client::builder()
+        .connect_timeout(connect_timeout)
+        .timeout(read_timeout)
+        .build()?;
+
+    let mut remaining = retry_for;
+    while !remaining.is_zero() {
+        let (resp, new_remaining) = http::post(
+            &client,
+            headers.clone(),
+            body.clone(),
+            read_timeout,
+            connect_timeout,
+            remaining,
+            url,
+        )
+        .await?;
+
+        match resp.status() {
+            StatusCode::CREATED | StatusCode::OK => {
+                tracing::info!("Provisioning-health report succeeded");
+                return Ok(());
+            }
+            StatusCode::TOO_MANY_REQUESTS => {
+                tracing::warn!("429 from wireserver, retrying…")
+            }
+            StatusCode::SERVICE_UNAVAILABLE => {
+                tracing::warn!("503 from wireserver, retrying…")
+            }
+            other => {
+                tracing::warn!("Non-retryable status, bailing out");
```

Comment:
can we add an info trace of every response from wireserver?  We can avoid this entire block with a simpler status code check